wsrep_membership_service.h

```diff
@@ -0,0 +1,128 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Wsrep-API is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with wsrep-API.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/** @file wsrep_membership_service.h
+ *
+ * This file defines interface for quering the immediate membership and
+ * members' states of the current configuration. The information is provided
+ * OUT OF ORDER to facilitate administrative tasks.
+ *
+ * The provider which is capable of using the service interface v1 must
+ * export the following functions.
+ *
+ * int wsrep_init_membership_service_v1(struct wsrep_membership_service_v1*)
+ * void wsrep_deinit_membership_service_v1()
+ *
+ * which can be probed by the application.
+ *
+ * The application must initialize the service via above init function
+ * before the provider is initialized via wsrep->init(). The deinit
+ * function must be called after the provider side resources have been
+ * released via wsrep->free().
+ */
+
+#ifndef WSREP_MEMBERSHIP_SERVICE_H
+#define WSREP_MEMBERSHIP_SERVICE_H
+
+#include "wsrep_api.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif /* __cplusplus */
+
+/**
+ * Member info structure extended to contain member state
+ */
+struct wsrep_member_info_ext
+{
+    struct wsrep_member_info base;
+    wsrep_seqno_t            last_committed;
+    enum wsrep_member_status status;
+};
+
+/**
+ * Extended membership structure
+ */
+struct wsrep_membership
+{
+    /**
+     * Epoch of the membership data (last time it was updated)
+     */
+    wsrep_uuid_t group_uuid;
+    /**
+     * Sequence number of the last received (not processed) action
+     */
+    wsrep_seqno_t last_received;
+    /**
+     * When the members' data was last updated
+     */
+    wsrep_seqno_t updated;
+    /**
+     * Current group state
+     */
+    enum wsrep_view_status state;
+    /**
+     * Number of members in the array
+     */
+    size_t num;
+    /**
+     * Membership array
+     */
+    struct wsrep_member_info_ext members[1];
+};
+
+/**
+ * Query membership
+ *
+ * @param wsrep      Provider handle
+ * @param membership Pointer to pointer to the memebrship structure.
+ *                   The structure is allocated by provider and must be freed
+ *                   by the caller.
+ * @return error code of the call
+ */
+typedef wsrep_status_t (*wsrep_get_membership_fn) (
+    wsrep_t* wsrep,
+    struct wsrep_membership** membership);
+
+/**
+ * Membership service struct.
+ * Returned by WSREP_MEMBERSHIP_SERVICE_INIT_FUNC_V1
+ */
+struct wsrep_membership_service_v1
```

Comment:
Looking at `wsrep_get_membership_fn` and this interface definition, it seems that the provider is responsible of allocating memory for the returned struct, and the application is responsible of releasing it. We already know that this does not work well in certain situations where application has been compiled with instrumented/custom allocator.

So my question is, should we extend this interface to have function to release memory, or should we introduce allocation service interface which could be used to delegate allocation to application (at least for the objects which lifetime crosses wsrep-API boundary).

Reply:
@temeo Yes, I see the issue.

I think, from the allocation management perspective allocation service interface is cleaner than the interface to release memory. I wonder if we simply can pass allocator pointer (`void* (*alloc)(size_t)`) as one of the arguments to `wsrep_get_membership_fn` (before we can pass it as a general callback in v27)

what you think?